### PR TITLE
[DEV 3566] Added indices on datasets

### PR DESCRIPTION
**Pull request by nonibansal**
1. Added offline and online indices on datasets with schema validation tests.
2. Removed timestamp.now() from code and added timestamp.utcnow().
3. Removed dataset.lookup from integration tests and added client.lookup.

### CODE REVIEW

File: fennel/datasets/datasets.py [hunk 1]
```diff
@@ -1700,6 +1707,184 @@ def sync_validation_for_pipelines(pipelines: List[Pipeline], ds_name: str):
...
+    offline: Optional[str] = DEFAULT_INDEX_OFFLINE,
+) -> Callable[[Dataset], Dataset]: ...
+
+
+@overload
+def index(obj: Dataset) -> Dataset: ...  # noqa: E704
+
+
+def index(
+    obj: Dataset = None,
```
Comment (aditya-nambiar) [inline fennel/datasets/datasets.py:1753]: Optional[Dataset]?

File: fennel/datasets/datasets.py [hunk 2]
```diff
@@ -1700,6 +1707,184 @@ def sync_validation_for_pipelines(pipelines: List[Pipeline], ds_name: str):
...
+    type: Literal["primary"] = DEFAULT_INDEX_TYPE,  # type: ignore
+    online: bool = DEFAULT_INDEX_ONLINE,
+    offline: Optional[str] = DEFAULT_INDEX_OFFLINE,
+) -> Union[Callable[[Dataset], Dataset], Dataset]:
+    """index decorator, defaults to a dataset with primary index, online set to true and forever lookup.
```
Comment (aditya-nambiar) [inline fennel/datasets/datasets.py:1758]: and offline index is forever retention ?

File: fennel/datasets/datasets.py [hunk 3]
```diff
@@ -1700,6 +1707,184 @@ def sync_validation_for_pipelines(pipelines: List[Pipeline], ds_name: str):
...
+            "Currently only forever retention is supported for Offline Index"
+        )
+        # duration = duration_proto.Duration()
+        # duration.FromTimedelta(index_obj.offline)
+        # offline = index_proto.OfflineIndex(
+        #     ds_version=obj.version,
+        #     ds_name=obj._name,
+        #     index_type=index_type,
+        #     duration=index_proto.IndexDuration(duration=duration),
+        # )
```
Comment (aditya-nambiar) [inline fennel/datasets/datasets.py:1884]: delete this code ?

File: fennel/datasets/test_invalid_indices.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+    assert (
+        str(e.value)
+        == "Extractor `extract` in `Featureset2` failed to run with error: Please define an online index on dataset : Dataset1. "
+    )
+
+
+@mock
+def test_invalid_dataset_offline_lookup(client):
+    """
+    This tests whether that a dataset doing offline lookup without online index is not allowed.
```
Comment (aditya-nambiar) [inline fennel/datasets/test_invalid_indices.py:155]: without offline index?

**Review by aditya-nambiar — APPROVED** [review]

File: fennel/datasets/datasets.py (hunk 2 above)
Comment (nonibansal) [reply fennel/datasets/datasets.py:1758]: yup

File: fennel/datasets/datasets.py (hunk 1 above)
Comment (nonibansal) [reply fennel/datasets/datasets.py:1753]: changed

File: fennel/datasets/datasets.py (hunk 3 above)
Comment (nonibansal) [reply fennel/datasets/datasets.py:1884]: removed

File: fennel/datasets/test_invalid_indices.py (hunk 4 above)
Comment (nonibansal) [reply fennel/datasets/test_invalid_indices.py:155]: changed

**Pull request merged**